packages/transactional-adapters/transactional-adapter-pg-promise/test/transactional-adapter-pg-promise.spec.ts

```diff
@@ -0,0 +1,205 @@
+} from '@nestjs-cls/transactional';
+import { Inject, Injectable, Module } from '@nestjs/common';
+import { Test, TestingModule } from '@nestjs/testing';
+import { ClsModule } from 'nestjs-cls';
+import { PgPromiseDbOrTx, TransactionalAdapterPgPromise } from '../src';
+import pgPromise from 'pg-promise';
+import { execSync } from 'node:child_process';
+
+const PG_PROMISE = 'PG_PROMISE';
+
+type UserRecord = { id: number; name: string; email: string };
+
+@Injectable()
+class UserRepository {
+    constructor(
+        private readonly txHost: TransactionHost<TransactionalAdapterPgPromise>,
+    ) {}
+
+    async getUserById(id: number) {
+        return this.txHost.tx.one<UserRecord>(
+            'SELECT * FROM public.user WHERE id = $1',
+            [id],
+        );
+    }
+
+    async createUser(name: string) {
+        const created = await this.txHost.tx.one<UserRecord>(
+            'INSERT INTO public.user (name, email) VALUES ($1, $2) RETURNING *',
+            [name, `${name}@email.com`],
+        );
+        return created;
+    }
+}
+
+@Injectable()
+class UserService {
+    constructor(
+        private readonly userRepository: UserRepository,
+        private readonly txHost: TransactionHost<TransactionalAdapterPgPromise>,
+        @Inject(PG_PROMISE)
+        private readonly db: PgPromiseDbOrTx,
+    ) {}
+
+    @Transactional()
+    async transactionWithDecorator() {
+        const r1 = await this.userRepository.createUser('John');
+        const r2 = await this.userRepository.getUserById(r1.id);
+        return { r1, r2 };
+    }
+
+    @Transactional<TransactionalAdapterPgPromise>({
+        isolationLevel: 'serializable', // FIXME
+    })
+    async transactionWithDecoratorWithOptions() {
+        const r1 = await this.userRepository.createUser('James');
+        const r2 = await this.db.oneOrNone<UserRecord>(
+            'SELECT * FROM public.user WHERE id = $1',
+            [r1.id],
+        );
+        const r3 = await this.userRepository.getUserById(r1.id);
+        return { r1, r2, r3 };
+    }
+
+    async transactionWithFunctionWrapper() {
+        return this.txHost.withTransaction(
+            {
+                isolationLevel: 'serializable',
+            },
+            async () => {
+                const r1 = await this.userRepository.createUser('Joe');
+                const r2 = await this.db.oneOrNone<UserRecord>(
+                    'SELECT * FROM public.user WHERE id = $1',
+                    [r1.id],
+                );
+                const r3 = await this.userRepository.getUserById(r1.id);
+                return { r1, r2, r3 };
+            },
+        );
+    }
+
+    @Transactional()
+    async transactionWithDecoratorError() {
+        await this.userRepository.createUser('Nobody');
+        throw new Error('Rollback');
+    }
+}
+
+const pgp = pgPromise();
+const db = pgp({
+    host: 'localhost',
+    user: 'postgres',
+    password: 'postgres',
+    database: 'postgres',
+});
```

Comment:
See it you can make it work with [`pg-mem`](https://github.com/oguimbal/pg-mem#-libraries-adapters). I would prefer this to docker compose, given we're not really testing how it works with the _database_ itself, just the database library.

Reply:
No luck, unfortunately.

There doesn't seem a concept of transaction in `pg-mem` and, at least from my tests, a in-progress transaction can see changes made by other transactions that haven't committed yet.

The author of `pg-mem` said [there's no support for concurrent transactions](https://github.com/oguimbal/pg-mem/issues/117#:~:text=pg%2Dmem%20does%20not%20support%20concurrent%20transactions).

Given this situation, are you okay with keeping docker compose?

Reply:
Well, that's a bummer.

Sure, go ahead and use docker-compose then. Ideally with a proper healthcheck and `docker-compose up -d --wait` so that it doesn't rely on arbitrary wait times.

And also please use some non-standard external port (like 5444) to expose the connection so that there's less chance of it colliding with other people's docker setups when running tests locally.

Reply:
All done!